python/src/com/jetbrains/python/psi/impl/PyTargetExpressionImpl.java

```diff
@@ -243,24 +244,37 @@ private static PyType getWithItemVariableType(TypeEvalContext context, PyWithIte
     final PyExpression expression = item.getExpression();
     if (expression != null) {
       final PyType exprType = context.getType(expression);
-      if (exprType instanceof PyClassType) {
-        final PyClass cls = ((PyClassType)exprType).getPyClass();
-        final PyFunction enter = cls.findMethodByName(PyNames.ENTER, true, null);
-        if (enter != null) {
-          final PyType enterType = enter.getCallType(expression, Collections.emptyMap(), context);
-          if (enterType != null) {
-            return enterType;
-          }
-          for (PyTypeProvider provider : Extensions.getExtensions(PyTypeProvider.EP_NAME)) {
-            PyType typeFromProvider = provider.getContextManagerVariableType(cls, expression, context);
-            if (typeFromProvider != null) {
-              return typeFromProvider;
-            }
-          }
-          // Guess the return type of __enter__
-          return PyUnionType.createWeakType(exprType);
+      if (exprType instanceof PyClass) {
+        final PyType enterType = getEnterTypeFromPyClass(context, expression, (PyClassType)exprType);
+        if (enterType != null)
+          return enterType;
+      }
+      else if (exprType instanceof PyUnionType) {
+        List<PyType> collect = ((PyUnionType)exprType).getMembers().stream().filter(PyClassType.class::isInstance)
```

Comment:
`StreamEx.of(...).select(PyClassType.class). ...` is shorter than `filter(...isInstance).map(...cast)`.